benchmark/operations/create_container.py

```diff
@@ -0,0 +1,319 @@
+import random
+from uuid import UUID, uuid4
+
+from pyrsistent import PClass, field
+from zope.interface import implementer
+
+from flocker.apiclient import MountedDataset
+from flocker.common import gather_deferreds, loop_until
+from flocker.control import DockerImage
+
+from .._interfaces import IProbe, IOperation
+
+
+class EmptyClusterError(Exception):
+    """
+    Exception indicating that the cluster contains no nodes.
+    """
+
+
+def create_dataset(
+    reactor, control_service, node_uuid, dataset_id, volume_size
+):
+    """
+    Create a dataset, then wait for it to be mounted.
+
+    :param IReactorTime reactor: Twisted Reactor.
+    :param IFlockerAPIV1Client control_service: Benchmark control
+        service.
+    :param UUID node_uuid: Node on which to create dataset.
+    :param UUID dataset_id: ID for created dataset.
+    :param int volume_size: Size of volume in bytes.
+    :return Deferred[DatasetState]: The state of the created dataset.
+    """
+
+    def dataset_mounted(expected):
+        """
+        Check whether a dataset has been created and mounted.
+
+        :param Dataset expected: A dataset configuration to match against the
+            results of ``list_datasets_state``.
+        :return Deferred[Optional[DatasetState]]: The matching dataset state,
+            or None if the expected dataset is not found.
+        """
+        d = control_service.list_datasets_state()
+
+        def dataset_matches(inspecting, expected):
+            return (
+                expected.dataset_id == inspecting.dataset_id and
+                expected.primary == inspecting.primary and
+                inspecting.path is not None
+            )
+
+        def find_match(existing_state):
+            for state in existing_state:
+                if dataset_matches(state, expected):
+                    return state
+            return None
+        d.addCallback(find_match)
+        return d
+
+    d = control_service.create_dataset(
+        primary=node_uuid,
+        maximum_size=volume_size,
+        dataset_id=dataset_id,
+    )
+
+    def loop_until_dataset_mounted(expected):
+        return loop_until(reactor, partial(dataset_mounted, expected))
+    d.addCallback(loop_until_dataset_mounted)
+
+    return d
+
+
+def create_container(
+    reactor, control_service, node_uuid, name, image, volumes=None
+):
+    """
+    Create a container, then wait for it to be running.
+
+    :param IReactorTime reactor: Twisted Reactor.
+    :param IFlockerAPIV1Client control_service: Benchmark control
+        service.
+    :param UUID node_uuid: Node on which to start the container.
+    :param unicode name: Name of the container.
+    :param DockerImage image: Docker image for the container.
+    :param Optional[Sequence[MountedDataset]] volumes: Volumes to attach
+        to the container.
+    :return Deferred[ContainerState]: The state of the created container.
+    """
+
+    def container_started(expected):
```

Comment:
:question: This is very similar to the dataset_mounted function above. Could at least
the machinery to do the loop over the states using a callable parameter to do the
matching be extracted?

In fact would it be useful to have a helper to do the loop_until part? It seems
like it is fairly common in the benchmarking code. I don't know if it would be useful
to have one specialised to take a state matcher and an operation that returns a list
of states and to loop until the callable return value is not None.